File: apidesign-3abae898011d/samples/individualsamples/src/org/apidesign/samples/XMMS.java
package org.apidesign.samples;

// BEGIN: xmms.java.api
public class XMMS {
    public void play() { doPlay(); }
    public void pause() { doPause(); }
    public void addToPlaylist(String file) { doAddToPlaylist(file); }
// FINISH: xmms.java.api

    // BEGIN: xmms.java.spi
    interface Playback {
        public void playback(byte[] data);
    }
    // END: xmms.java.spi
    public void registerPlayback(Playback callback) {
        this.callback = callback;
    }
    private Playback callback;
    
    private void doAddToPlaylist(String file) {
    }

    private void doPause() {
    }

    private void doPlay() {
        callback.playback("Play!".getBytes());
    }
    
    
    
    //
    // Usage
    // 
    
    public static void main(String[] args) {
        XMMS xmms = new XMMS();
        
        // BEGIN: xmms.java.use
        class MyCallbackPrints implements XMMS.Playback {
            public void playback(byte[] data) {
                System.out.println(new String(data));
            }
        }
        xmms.registerPlayback(new MyCallbackPrints());
        // END: xmms.java.use
    }

}

File: apidesign-3abae898011d/samples/individualsamples/src/org/apidesign/samples/xmms.c

#include "stdio.h"

// BEGIN: xmms.spi
void xmms_register_playback(void (*f)(char*));
// END: xmms.spi

// BEGIN: xmms.api
void xmms_play();
void xmms_pause();
void xmms_add_to_list(char *);
// END: xmms.api


//
// implementation
//

void none() {
}

void (*callback)(char*) = none;
void xmms_register_playback(void (*f)(char*)) {
    callback = f;
}

void xmms_play() {
    callback("Play!");
}

void xmms_pause() {
}

//
// test case
//

// BEGIN: xmms.your.playback
void my_playback_prints(char* text) {
  printf("%s\n", text);
}
// END: xmms.your.playback

int main(char** argsv) {
    xmms_register_playback(my_playback_prints);
    xmms_play();
    xmms_pause();
}

